### ISSUE TITLE
Reading Pointer bytes as Integers

### ISSUE THREAD

**Issue by chorman0773** [rust-lang/unsafe-code-guidelines#547]
This came up in https://github.com/rust-lang/reference/pull/1664. I wanted to ask what T-opsem thinks about the behaviour of reading pointer bytes as integer types (or as `char`/`bool`/etc.). 

As far as I can tell, there are two "sensible" behaviours, given that integers themselves do no carry provenance:
* The pointer fragment is ignored,
* Decoding error (thus undefined behaviour).

Given provenance monotonicity, which would be violated by the decoding error, it seems like the best option is that the fragments are ignored. Is there anything missed here? If not, can we get a formal sign off on this behaviour.

Note that I'm only considering the runtime behaviour, which can be a point against adopting the behaviour. Given that it's impossible to get the address of certain pointers in const-eval, it does need to be undefined behaviour (or otherwise an error) to read pointer bytes (to at least symbolic allocations) as integer types.

**Comment by saethlin** [comment]
> Given that it's impossible to get the address of certain pointers

Which pointers?

**Comment by chorman0773** [comment]
I failed to clarify that. It was referring to the consteval AM, where allocations that exist outside of the particular constant evaluation (what I call symbolic pointers) can't be assigned an address.

**Comment by RalfJung** [comment]
Const-eval can't assign an address to *any* allocation, "inside" or "outside". (Not sure what you mean with that distinction.)

**Comment by chorman0773** [comment]
I think it could if the allocation is solely within the AM, since promoted values are allocated after the appropriate consteval AM is finished, and no other allocation can be "leaked" out of a constant expression, and thus never be observed otherwise.

**Comment by RalfJung** [comment]
>  since promoted values are allocated after the appropriate consteval AM is finished

That doesn't work, const-eval has access to plenty of things that must have their address fixed already during const-eval time. The easiest is `static`s.
```rust
static S: i32 = 0;
const C: usize = {
  let x = 0;
  &x as *const i32 as usize
};

fn main() {
  println!("{} {}", C, &S as *const i32 as usize);
}
```
This must never print the same address.

**Comment by RalfJung** [comment]
Anyway that sub-discussion seems off-topic here, please move it to Zulip. And please update the issue description to clarify that "certain pointers" refers to const-eval.

**Comment by chorman0773** [comment]
I suppose the *third* alternative that should be addressed is that the read exposes the pointer bytes, but I don't like that suggestion (and I recall few people did), as it means that reads can result in a side effect, and such reads as an integer type can never be elided.

Is there any other alternative I'm missing?

**Comment by RalfJung** [comment]
Yeah I definitely don't like that suggestion, it pessimizes optimization too much. It is worth mentioning that that third alternative is basically what PNVI-ae-udi mandates for C. I am curious if compilers will actually implement that, though.

**Comment by RalfJung** [comment]
> Note that I'm only considering the runtime behaviour, which can be a point against adopting the behaviour. Given that it's impossible to get the address of certain pointers in const-eval, it does need to be undefined behaviour (or otherwise an error) to read pointer bytes (to at least symbolic allocations) as integer types.

We could characterize this as a "unsupported in const-eval" error rather than a UB error. (Internally in rustc this is already what we do, `ReadPointerAsInt` is a variant of `UnsupportedOpInfo`. However we don't clearly distinguish those cases in the error message AFAIK, and we do call this UB in the `transmute` docs.)

**Comment by RalfJung** [comment]
There's another aspect of provenance that we haven't officially decided yet and that is implicitly excluded by the current wording in https://github.com/rust-lang/reference/pull/1664: do the individual bytes in a pointer "remember" where in the pointer they are, and have to be put back in the same order? Some formal models require this, and if we ever allow "taking apart" the bytes of a pointer in const-eval we'll also have to require this, but for runtime semantics we could decide either way.

The one example of code that I am aware of that breaks this requirement is XOR linked lists, which can be implemented in the semantics sketched in MiniRust right now but can't be implemented if bytes with provenance remember their position in the pointer. That's not exactly realistic code, but it is somewhat satisfying that (on architectures where pointers have at least 2 bytes), XOR linked lists can be implemented.

The main upside of requiring the same bytes in the same order is that it rules out pointer crimes like XOR linked lists so if there's some unexpected interactions there, we'd not be affected. I am not aware of an optimization that would benefit from this UB, it's mostly a case of "ruling out some rather cursed programs to avoid locking ourselves into an unexpected corner". In some sense the model becomes a bit simpler since we can just say, pointer bytes must be put back together in the same order they started out as before they can be treated as a pointer again, but the actual op.sem would become more complicated because of the extra bookkeeping is required to enforce this.

**Comment by CatsAreFluffy** [comment]
This sounds a lot like #286.

**Comment by RalfJung** [comment]
True, those are discussing the same thing.

**Comment by RalfJung** [comment]
What is hard to evaluate is whether there isn't a better set of atomic building blocks that better captures the relevant properties. Provenance is a relevant concept so it seems reasonable to expect it to be reflected somehow.

**Comment by joshlf** [comment]
Yeah, we do intend to reflect provenance. But we'd prefer to reflect that "ptr-to-int is valid but strips provenance" rather than have to not support ptr-to-int because it's UB.

**Comment by RalfJung** [comment]
Closing in favor of https://github.com/rust-lang/unsafe-code-guidelines/issues/286.
